pkg/master/controller/crdregistration/crdregistration_controller.go

```diff
@@ -208,19 +224,44 @@ func (c *crdRegistrationController) handleVersionUpdate(groupVersion schema.Grou
 				continue
 			}
 
-			c.apiServiceRegistration.AddAPIServiceToSync(&apiregistration.APIService{
+			apiService := &apiregistration.APIService{
 				ObjectMeta: metav1.ObjectMeta{Name: apiServiceName},
 				Spec: apiregistration.APIServiceSpec{
 					Group:                groupVersion.Group,
 					Version:              groupVersion.Version,
 					GroupPriorityMinimum: 1000, // CRDs should have relatively low priority
 					VersionPriority:      100,  // CRDs will be sorted by kube-like versions like any other APIService with the same VersionPriority
 				},
-			})
+			}
+			c.apiServiceRegistration.AddAPIServiceToSync(apiService)
+			// Serve CRD OpenAPI spec if we have non-nil openAPIAggregationManager
+			if c.openAPIAggregationManager != nil {
+				// Convert internal CustomResourceDefinitionSpec to versioned CustomResourceDefinitionSpec
+				versionedSpec := new(v1beta1.CustomResourceDefinitionSpec)
+				if err := v1beta1.Convert_apiextensions_CustomResourceDefinitionSpec_To_v1beta1_CustomResourceDefinitionSpec(&crd.Spec, versionedSpec, nil); err != nil {
+					return err
+				}
+				crdspec, etag, err := apiextensionsopenapi.CustomResourceDefinitionOpenAPISpec(versionedSpec)
```

Comment:
This is "the one call into apiextensions-apiserver"?

Wondering, could we push the whole openapi spec generation for all CRDs into apiextensions-apiserver and then only merge it into the kube-apiserver openapi spec?

Think of apiextensions-apiserver running standalone. The same code path should be used in that case.

Reply:
I think there are three parts of the problem:

1. Generating the spec for a new added/updated CRD: 

This can happen at any place along the CRD adding/updating path. Currently in this PR we put it at where the KAS discovery registration happens, but we should be able to move it to `k8s.io/apiextensions-apiserver` (maybe in one of the [CRD controllers](https://github.com/kubernetes/kubernetes/blob/c0be4cc09c5a685331cc310273cfdd633eac54ac/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go#L196-L199)).

2. Serving (aggregating) the spec to apiextensions apiserver:

This requires writing to the OpenAPIService that gets hooked to `apiExtensionsServer.GenericAPIServer.Handler`. Currently the [OpenAPIService](https://github.com/kubernetes/kubernetes/blob/c0be4cc09c5a685331cc310273cfdd633eac54ac/cmd/kube-apiserver/app/server.go#L193-L194) gets created after [apiextensions apiserver](https://github.com/kubernetes/kubernetes/blob/c0be4cc09c5a685331cc310273cfdd633eac54ac/cmd/kube-apiserver/app/server.go#L179) gets created. 

To use the same code path for standalone apiextensions apiserver, code in `k8s.io/apiextensions-apiserver` needs to have access to the OpenAPIService. I'm thinking how can we pass it to one of the CRD controllers. @sttts Any suggestions?

3. Aggregating the apiextensions apiserver spec to aggregator apiserver spec:

This should happen in staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go